packages/ai-chat/src/common/chat-agents.ts

```diff
@@ -0,0 +1,384 @@
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+// Partially copied from https://github.com/microsoft/vscode/blob/a2cab7255c0df424027be05d58e1b7b941f4ea60/src/vs/workbench/contrib/chat/common/chatAgents.ts
+
+import {
+    CommunicationRecordingService,
+    getTextOfResponse,
+    LanguageModel,
+    LanguageModelRequirement,
+    LanguageModelResponse,
+    PromptService,
+    ResolvedPromptTemplate,
+    ToolRequest,
+} from '@theia/ai-core';
+import {
+    Agent,
+    isLanguageModelStreamResponse,
+    isLanguageModelTextResponse,
+    LanguageModelRegistry,
+    LanguageModelStreamResponsePart,
+    MessageActor,
+    PromptTemplate
+} from '@theia/ai-core/lib/common';
+import { CancellationToken, CancellationTokenSource, ILogger, isArray } from '@theia/core';
+import { inject, injectable } from '@theia/core/shared/inversify';
+import { ChatAgentService } from './chat-agent-service';
+import {
+    ChatModel,
+    ChatRequestModel,
+    ChatRequestModelImpl,
+    ChatResponseContent,
+    CodeChatResponseContentImpl,
+    ErrorResponseContentImpl,
+    MarkdownChatResponseContentImpl,
+    ToolCallResponseContentImpl
+} from './chat-model';
+
+export interface ChatMessage {
+    actor: MessageActor;
+    type: 'text';
+    query: string;
+}
+
+export interface SystemMessage {
+    text: string;
+    /** All functions references in the system message. */
+    functionDescriptions?: Map<string, ToolRequest<object>>;
+}
+export namespace SystemMessage {
+    export function fromResolvedPromptTemplate(resolvedPrompt: ResolvedPromptTemplate): SystemMessage {
+        return {
+            text: resolvedPrompt.text,
+            functionDescriptions: resolvedPrompt.functionDescriptions
+        };
+    }
+}
+
+export enum ChatAgentLocation {
+    Panel = 'panel',
+    Terminal = 'terminal',
+    Notebook = 'notebook',
+    Editor = 'editor'
+}
+
+export namespace ChatAgentLocation {
+    export const ALL: ChatAgentLocation[] = [ChatAgentLocation.Panel, ChatAgentLocation.Terminal, ChatAgentLocation.Notebook, ChatAgentLocation.Editor];
+
+    export function fromRaw(value: string): ChatAgentLocation {
+        switch (value) {
+            case 'panel': return ChatAgentLocation.Panel;
+            case 'terminal': return ChatAgentLocation.Terminal;
+            case 'notebook': return ChatAgentLocation.Notebook;
+            case 'editor': return ChatAgentLocation.Editor;
+        }
+        return ChatAgentLocation.Panel;
+    }
+}
+
+export interface ChatAgentData extends Agent {
+    locations: ChatAgentLocation[];
+    iconClass?: string;
+}
+
+export const ChatAgent = Symbol('ChatAgent');
+export interface ChatAgent extends ChatAgentData {
+    invoke(request: ChatRequestModelImpl, chatAgentService?: ChatAgentService): Promise<void>;
```

Comment:
The `ChatRequestModel` is the interface exposed to the "outside" world. For example any interested code can get a hold of all current chat sessions via the `ChatService`.

The `XYZImpl` classes expose far more, especially methods on these classes to modify them. They need knowledge of the inner workings of the classes and should only exposed to the ones which actually work with them. Agents are the main work force in the AI design, so they get handed over the `Impl` classes as they need to use the richer interface.

VS Code uses the same design pattern in their code, it just looks different because of the conventions there, i.e. `IChatRequestModel / ChatRequestModel` vs `ChatRequestModel / ChatRequestModelImpl`.

Reply:
If there are two different interfaces, then let's make two different interfaces. Otherwise, it's unclear what protocol is available to contributors and which isn't. But anyway, I really don't see any interface that the implementation class exposes which the interface does not. There's a bunch of protected variables, but those are only available to extenders of the class.

Reply:
The important difference between `ChatRequestModel` and `ChatRequestModelImpl` here is that the `ChatRequestModelImpl` links to `ChatResponseModelImpl` instead of `ChatResponseModel`.

`ChatResponseModelImpl` has a lot of additional functionality to manipulate the response which one major part of agent functoniality. This would not be available to the agent then.

We could introduce additional interfaces for this instead of relying on the `Impl` classes. Any preference on their naming? e.g. `InternalChatRequestModel`, `AgentChatRequestModel`? If it's fine for you I would handle this in the follow up.

Reply:
I'm not really sure why we need two different interfaces at all. Security is not a concern here, since we're talking about compile time API's anyway. If you're saying "outside world", you're still referring to third-party contributed chat agents that can manipulate the response object. Nothing is keeping them from taking a reference to the mutable obejct and sharing it. So it's not "internal" vs. "external", but we're talking about different roles in the system. I'm not sure that's a decision we need to make for consumers of the API.
Since it's not internal/external, the name would be more like "MutableChatRequest".

We have a kind of similar use case in the Test API: from the point of view of the contributor, the API is basically write-only: you can register a test controller and you can add test items to your own controller, but you can never get a list of all test controllers from the system. Maybe we're in that kind of scenario here?

Reply:
But we can talk about this maybe offline and address issues we identify in follow-ups.

Reply:
The difference is between the clients of the Chat API
- which just interact with it, e.g. the Chat UI which is just rendering the current sessions, sending a request etc, AND
- the producing side, i.e. the ChatService and agents which produce these models and modify them during runtime

The first clients are the ones which shall only work on the public API. For them all fields are "readonly", i.e. they are not allowed to modify them. What they can do is listening to changes and then recheck the fields. The other "clients" of the API are the producers, i.e. the ChatService and agents who are allowed and are intended to modify the fields, e.g. append some response characters when received by the LLM, eagerly analyzing the response and producing corresponding response parts etc.

So yes, I think you summarized it correctly with the different roles in the system. I like the `Mutable` naming suggestion. If it's fine for you I'll suggest it for the follow up